### PR TITLE
Add proposal doc for integration between Harbor and Ali Dragonfly

### PR DESCRIPTION
**Pull request by steven-zou**
**Add proposal doc for integration between Harbor and Ali Dragonfly**
Signed-off-by: Steven Zou <szou@vmware.com>

### CODE REVIEW

File: proposals/new/integrate-with-dragonfly_proposal.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  * API endpoint
+  * API access token or required credentials
+  * Possible calling policies or automation rules etc.
+
+* The integrated configurations can be verified to make sure the connection between the two systems is not broken by testing or ''dry run" etc.
+
+* The images are produced by CI/CD pipeline or any other ways and pushed to the Harbor registry. The newly pushed images can be marked with labels automatically or manually. In addition, the admin of the registry can also scan the images to make sure it's secure. Of course, the admins can do any other management work if they want.
+
+* The admin of the registry can select any ready image to **promote** it to the supervise node of Dragonfly P2P network for the upcoming image pulling requests to improve the distribution performance. The **promote** action can be triggered by clicking button or auto-triggered by pre-configured rules/policies (_If match some conditions, then promote it_).
```
Comment (clouderati) [inline proposals/new/integrate-with-dragonfly_proposal.md]: This implies that images are not automatically distributed via Dragonfly. I think this is ok for a MVP, but long-term I think it makes sense to:

* Have specific projects backed by Dragonfly such that any and all images uploaded to said projects are automatically distributed
* Have system-wide distribution (on all projects) if configured as such

Thoughts?

**Comment by clouderati** [comment]
Really happy to see this moving forward; super exciting prospects here!

I believe, however, that this opens a pandora's box of questions regarding a number of things:

* Two registries, `harborA` and `harborB` are part of the Dragonfly P2P network. An image is pushed to `harborA` and immediately promoted and distributed transparently by Dragonfly. What happens to the metadata associated with that image, such as:
  * RBAC – the project rules that govern access to an image in `harborA` will not immediately get replicated to `harborB` and the relevant project(s)
  * Vulnerability information – will the image have to be scanned again when it hits `harborB`?
  * Labels – ditto above

Overall I think this proposal is OK, but this is an opportune time to really think about Harbor on a macro-scale and how we can add truly distributed functionality. Proposals this sweeping would, ermmm, belong in another proposal, obviously. :smile:

**Comment by steven-zou** [comment]
> Hi @steven-zou – any updates on this based on your discussion with the Dragonfly team? Can't wait to see this functionality land!

@clouderati 

 From the harbor side, overall progress (both UI and API) is 60%. It should be ready before KubeCon NA

**Comment by clouderati** [comment]
Wahoo! Great, thanks @steven-zou.

**Comment by steven-zou** [comment]
Prototype is here:

https://github.com/steven-zou/harbor/tree/integrate_with_dragonfly

File: proposals/new/integrate-with-dragonfly_proposal.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+### Basic Workflow
+
+![harbor dragonfly](https://user-images.githubusercontent.com/5753287/42671176-576f857a-8691-11e8-802e-7f39e836cffa.png)
+
+### Architecture
+
+An architecture design based on the above draft idea:
+![dragonfly h](https://user-images.githubusercontent.com/5753287/43447108-4ba61ac4-94dd-11e8-9072-0c143b58e70d.png)
+
+The components with light blue background are the new things need to be implemented.
```
[the remark this answers is not in the document]
Comment (steven-zou) [reply proposals/new/integrate-with-dragonfly_proposal.md]: @cd1989 

No, they're running in the harbor-core service.

**Comment by datavisoryushuzhang** [comment]
**Awesome !!!**
When could this prototype be released? @steven-zou

**Comment by steven-zou** [comment]
We heard lots of voices from the community that they're looking forward to such a great feature. Meanwhile, several community contributors expressed their interests in the related work of delivering this feature. We'll create a new Harbor workgroup focusing on the topic of P2P distribution and try to deliver a formal feature in the feature!

@cd1989 will provide some help doing coordinations to this group as co-coordinator. Thank @cd1989 for the support!

**Comment by steven-zou** [comment]
Meeting recording for the discussion with the community on June 21st: https://VMware.zoom.us/recording/share/wbdM90X5ALJeJIfdeWOQ2gGRCdddipMNpjrkjxXQ8HywIumekTziMw

**Comment by AkihiroSuda** [comment]
What's current status? @steven-zou @cd1989

**Comment by steven-zou** [comment]
New proposal is drafted to reflect the latest discussion and new design. Please refer to https://github.com/goharbor/community/pull/142. I'll archive this one.

**Pull request closed without merging**